### PR TITLE
Adding @PathPrefix annotation

### PR DESCRIPTION
**Pull request by sivaalli**
Adding @PathPrefix annotation for a user to declare a prefix path that a service object handles.
Motivation :  https://github.com/line/armeria/issues/2031

This PR allows a user to declare a 

```
ServerBuilder sb = new ServerBuilder();
sb.annotatedService(new ForBarService());
```
and 
```
@PathPrefix("/foo")
public class FooBarService {
    @Get("/bar")
    public Object barGet() { ... }
}
```

But this PR does **not** allow something like this:

```
@PathPrefix("/foo")
public class FooBarService {
    @Get //without path
    public Object barGet() { ... }
}
```
Because a path has to be specified either in `@Path` or in a http method annotation(`@Get`) or an exception is thrown. But I think this is nice to have feature. I just want to get feedback on if this is the right direction I'm going and this is what expected. Because the issue did not directly talk about removing path to be specified in `@Path` or http method annotations. Please let me know if you want to see this.
Also, I will be doing documentation changes after this is fully done.

### CODE REVIEW

**Comment by trustin** [comment]
@sivaalli, thank you so much for your first contribution! :clap: 

I agree that the following should not be allowed:

```java
@PathPrefix("/foo")
public class FooBarService {
    @Get //without path
    public Object barGet() { ... }
}
```

.. while the following should be allowed:

```java
@PathPrefix("/foo")
public class FooBarService {
    @Get("/bar") // mapped at "/foo/bar"
    public Object barGet() { ... }
}
```

Sounds good?

**Comment by trustin** [comment]
Oops. Updated the above comment.

**Comment by sivaalli** [comment]
Thank you @trustin  I'm really exited to contribute and looking forward to many more contributions as well. Also really appreciate your work on netty and armeria.

In my personal opinion, I can see a use case where a user wants to map a service to a prefix like `@PathPrefix("/users")` and have all the service methods handle different Http methods for ex:

```
    @PathPrefix("/users")
    public lass FooBarService {

        @Post
        public HttpResponse adduser(User user) { ... }

        @Get
        public HttpResponse getAllUsers() { ... }
    }
```
What do you think? Or am I missing something here.

**Comment by ikhoon** [comment]
@sivaalli Thanks for contributing! `@PathPrefix` annotation looks good to me. By the way, are you using Armeria in production? Could you share your experience?😁

**Comment by sivaalli** [comment]
> @sivaalli Thanks for contributing! `@PathPrefix` annotation looks good to me. By the way, are you using Armeria in production? Could you share your experience?😁

I work at Salesforce and a sub-cloud called Social, and we are currently on a monolith stack, but we are in talks of moving from JSON over REST to gRPC. I would definitely mention about Armeria when are are implementing it.

**Comment by sivaalli** [comment]
Hi Again, If this is **not** expected behaviour
```
@PathPrefix("/foo")
public class FooBarService {
    @Get //without path
    public Object barGet() { ... }
}
```
Then I think this PR is ready for review. Please let me know If I can start making Documentation changes or any other changes with respect to this PR. :). Thanks @heowc for letting me contribute to this issue.

**Comment by heowc** [comment]
Very nice! 👍

File: core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactory.java [hunk 1]
```diff
@@ -759,6 +761,15 @@ static String findDescription(AnnotatedElement annotatedElement) {
         return null;
     }
 
+    /**
+     * Return the path prefix to use. If there is {@link PathPrefix} annotation on the class
+     * then the value from the annotation is used, else falls back to pathPrefix passed through
+     * method pram.
+     */
+    private static String selectPathPrefix(Class<?> clazz, String pathPrefix) {
+        return findFirst(clazz, PathPrefix.class).map(PathPrefix::value).orElse(pathPrefix);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactory.java]: How about concatenating `pathPrefix` and `PathPrefix.value()`? For example:

```java
ServerBuilder sb = new ServerBuilder();
sb.annotatedService("/foo", new MyService());

@PathPrefix("/bar")
public class MyService { ... }
```

It seems natural to me that `MyService` is bound under `/foo/bar`. What do you think?

File: core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+package com.linecorp.armeria.server.annotation;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Annotation for mapping dynamic web requests onto specific class.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java]: Could you rephrase this to reflect what this class does more correctly?

File: core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactoryTest.java [hunk 3]
```diff
@@ -122,6 +131,35 @@ public void ofUserDefinedRepeatableDecorator() throws NoSuchMethodException {
         assertThat(udd2.value()).isEqualTo(2);
     }
 
+    @org.junit.jupiter.api.Test
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactoryTest.java]: I wouldn't recommend mixing JUnit 4 and 5 annotations. If you are not going to migrate this test case to JUnit 5, I recommend sticking to JUnit 4.

File: core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactoryTest.java [hunk 4]
```diff
@@ -122,6 +131,35 @@ public void ofUserDefinedRepeatableDecorator() throws NoSuchMethodException {
...
+                                           .flatMap(Collection::stream)
+                                           .collect(Collectors.toList());
+
+        assertThat(paths).allMatch(path -> path.startsWith(HOME_PATH_PREFIX + '/'));
+    }
+
+    @org.junit.jupiter.api.Test
+    public void testFindAnnotatedServiceElementsWithoutPathPrefixAnnotation() {
+        final Object serviceObject = new ServiceObject();
+        final String root = "/home";
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactoryTest.java]: Could be replaced with `HOME_PATH_PREFIX`?

File: core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactory.java (hunk 1 above)
Comment (heowc) [reply core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactory.java]: I agree. Otherwise, I think it can confuse the user. 👍

Comment (sivaalli) [reply core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactory.java]: Sounds good. Will make those changes. Thanks for review:)

File: core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java (hunk 2 above)
Comment (sivaalli) [reply core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java]: Sure. Thanks

File: core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactoryTest.java (hunk 3 above)
Comment (sivaalli) [reply core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactoryTest.java]: Sure. Will use junit4 annotations.

File: core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactoryTest.java (hunk 4 above)
Comment (sivaalli) [reply core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactoryTest.java]: Sure.

File: core/src/main/java/com/linecorp/armeria/server/RouteBuilder.java [hunk 5]
```diff
@@ -173,7 +173,6 @@ public RouteBuilder regex(Pattern regex) {
      * @see #path(String)
      */
     public RouteBuilder pathWithPrefix(String prefix, String pathPattern) {
-        prefix = ensureAbsolutePath(prefix, "prefix");
```
Comment (sivaalli) [inline core/src/main/java/com/linecorp/armeria/server/RouteBuilder.java]: This validation is done in `computePathPrefix` method

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/server/RouteBuilder.java]: Without this, a user will get an exception later than when he or she called this method. I think it gives a user better feed back when an exception is triggered as soon as a bad input is given.

File: core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import com.linecorp.armeria.server.ServerBuilder;
+
+/**
+ * Annotation that can be used on a class as a path prefix for all the
+ * methods that handle http request. For ex:
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java]: For example

File: core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+ * <pre>
+ * {@code @PathPrefix("/b")}
+ * public class MyService {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java]: You can do this in our Javadoc:

```java
 * <pre>{@code
 * > @PathPrefix("/b")
 * > public class MyService {
 * >     @Get("/c")
 * >     public HttpResponse foo() { ... }
 * > }
 * }</pre>
```

File: core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+ *     {@code @Get("/c")}
+ *     public HttpResponse foo() {...}
+ * }
+ * </pre>
+ * And then can be registered to {@link ServerBuilder} like this
+ * <pre>{@code
+ * ServerBuilder sb = new ServerBuilder();
+ * sb.annotatedService("/a", new MyService());
+ * }</pre>
+ * In this case <b>foo</b> methods handles a request that matches path <b>a/b/c</b>
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java]:
```
 * }</pre>
 *
 * <p>In this case, the {@code foo()} method handles a request that matches the path 
 * {@code /a/b/c}.</p>
```

File: core/src/test/java/com/linecorp/armeria/internal/ArmeriaHttpUtilTest.java [hunk 9]
```diff
@@ -71,6 +71,7 @@ void testConcatPaths() throws Exception {
 
         assertThat(concatPaths("/", "a")).isEqualTo("/a");
         assertThat(concatPaths("/", "/a")).isEqualTo("/a");
+        assertThat(concatPaths("/", "/")).isEqualTo("/");
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/internal/ArmeriaHttpUtilTest.java]: Thanks for adding more test cases.

**Comment by trustin** [comment]
Removed `[WIP]` :smile:

File: core/src/main/java/com/linecorp/armeria/server/RouteBuilder.java (hunk 5 above)
Comment (sivaalli) [reply core/src/main/java/com/linecorp/armeria/server/RouteBuilder.java]: Added this back.

File: core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java (hunk 6 above)
Comment (sivaalli) [reply core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java]: Corrected this :)

File: core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java (hunk 7 above)
Comment (sivaalli) [reply core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java]: Used this block. Thank you.

File: core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java (hunk 8 above)
Comment (sivaalli) [reply core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java]: Corrected this :)

File: core/src/test/java/com/linecorp/armeria/internal/ArmeriaHttpUtilTest.java (hunk 9 above)
Comment (sivaalli) [reply core/src/test/java/com/linecorp/armeria/internal/ArmeriaHttpUtilTest.java]: No Problem :)

**Comment by sivaalli** [comment]
Build failure is unrelated to this PR.

File: core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2016 LINE Corporation
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java]: nit: 2019

File: core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+ * >     public HttpResponse foo() { ... }
+ * > }
+ * }</pre>
+ * And then can be registered to {@link ServerBuilder} like this
+ * <pre>{@code
+ * ServerBuilder sb = new ServerBuilder();
+ * sb.annotatedService("/a", new MyService());
+ * }</pre>
+ *
+ * <p>In this case {@code foo()} methods handles a request that matches path {@code /a/b/c}</p>
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java]: nit: methods -> method

File: core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactoryTest.java [hunk 12]
```diff
@@ -122,6 +132,35 @@ public void ofUserDefinedRepeatableDecorator() throws NoSuchMethodException {
...
     }
 
+    @Test
+    public void testFindAnnotatedServiceElementsWithPathPrefixAnnotation() {
+        final Object object = new PathPrefixServiceObject();
+        final List<AnnotatedHttpServiceElement> elements = find("/", object, new ArrayList<>());
+
+        final List<String> paths = elements.stream()
+                                           .map(AnnotatedHttpServiceElement::route)
+                                           .map(Route::paths)
```
Comment (minwoox) [inline core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactoryTest.java]: This returns the list which has two same paths for one `route`. How about just getting the first element here?
```java
final List<String> paths = elements.stream()
                                   .map(AnnotatedHttpServiceElement::route)
                                   .map(route -> route.paths().get(0))
                                   .collect(Collectors.toList());
assertThat(paths).containsExactly("/home/hello", "/home/");
...

File: core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactoryTest.java [hunk 13]
```diff
@@ -122,6 +132,35 @@ public void ofUserDefinedRepeatableDecorator() throws NoSuchMethodException {
...
+
+    @Test
+    public void testFindAnnotatedServiceElementsWithoutPathPrefixAnnotation() {
+        final Object serviceObject = new ServiceObject();
+        final List<AnnotatedHttpServiceElement> elements = find(HOME_PATH_PREFIX, serviceObject,
+                                                                new ArrayList<>());
+
+        final List<String> paths = elements.stream()
+                                           .map(AnnotatedHttpServiceElement::route)
+                                           .map(Route::paths)
```
Comment (minwoox) [inline core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactoryTest.java]: ditto

File: core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactory.java [hunk 14]
```diff
@@ -759,6 +763,22 @@ static String findDescription(AnnotatedElement annotatedElement) {
         return null;
     }
 
+    /**
+     * Return the path prefix to use. If there is {@link PathPrefix} annotation on the class
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactory.java]: nit: Returns

File: core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactory.java [hunk 15]
```diff
@@ -759,6 +763,22 @@ static String findDescription(AnnotatedElement annotatedElement) {
...
+     * then path prefix is computed by concatenating pathPrefix and value from annotation else
+     * return pathPrefix
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactory.java]: nit: missing ending `.`

File: core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactoryTest.java [hunk 16]
```diff
@@ -224,4 +259,31 @@ public String userDefinedRepeatableDecorator() {
             return "";
         }
     }
+
+    @PathPrefix(HOME_PATH_PREFIX)
+    static class PathPrefixServiceObject {
+
+        @Get("/hello")
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactoryTest.java]: Could you add test cases for verifying exception with `@Get`, `@Post`? 😀

File: core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java (hunk 7 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/server/annotation/PathPrefix.java]: Javadoc will show `>` but we have some JavaScript that strips it out and applies syntax highlighting, so please keep using `>`. :smile:

File: core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactoryTest.java (hunk 16 above)
Comment (sivaalli) [reply core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactoryTest.java]: Yup I will do that. Thanks for suggestion.

File: core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactory.java [hunk 17]
```diff
@@ -255,7 +258,8 @@ private static HttpStatus defaultResponseStatus(Optional<HttpStatus> defaultResp
      * Returns an {@link AnnotatedHttpService} instance defined to {@code method} of {@code object} using
      * {@link Path} annotation.
      */
-    private static AnnotatedHttpServiceElement create(String pathPrefix, Object object, Method method,
+    @VisibleForTesting
+    static AnnotatedHttpServiceElement create(String pathPrefix, Object object, Method method,
                                                       List<ExceptionHandlerFunction> baseExceptionHandlers,
                                                       List<RequestConverterFunction> baseRequestConverters,
                                                       List<ResponseConverterFunction> baseResponseConverters) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactory.java]: Indentation

File: core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactoryTest.java [hunk 18]
```diff
@@ -122,6 +141,46 @@ public void ofUserDefinedRepeatableDecorator() throws NoSuchMethodException {
...
+        final List<String> paths = elements.stream()
+                                           .map(AnnotatedHttpServiceElement::route)
+                                           .map(route -> route.paths().get(0))
+                                           .collect(Collectors.toList());
+
+        assertThat(paths).containsExactlyInAnyOrder(HOME_PATH_PREFIX + "/hello", HOME_PATH_PREFIX + '/');
+    }
+
+    @Test
+    public void testCreateAnnotatedServiceElementWithoutExplicitPathOnMethod() {
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactoryTest.java]: Nice!

**Comment by sivaalli** [comment]
Thank you @trustin @minwoox @ikhoon for your support. I look forward to contribute more to Armeria.
